js/src/components/paid-ads/asset-group/texts-editor.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
+import GridiconCrossSmall from 'gridicons/dist/cross-small';
+
+/**
+ * Internal dependencies
+ */
+import AppButton from '.~/components/app-button';
+import AppInputControl from '.~/components/app-input-control';
+import AddAssetItemButton from './add-asset-item-button';
+import './texts-editor.scss';
+
+/**
+ * Renders a list of text inputs for managing the single type of asset texts.
+ *
+ * @param {Object} props React props.
+ * @param {string[]} [props.initialTexts=[]] Initial texts.
+ * @param {number} [props.minNumberOfTexts=0] Minimum number of texts.
+ * @param {number} [props.maxNumberOfTexts=0] Maximum number of texts.
+ * @param {number|number[]} props.maxCharacterCounts Maximum number of characters for each text. If the limits are the same, a single number can be used instead of an array.
+ * @param {string} props.addButtonText Text for the button to add a new text input.
+ * @param {string} [props.placeholder] Placeholder text.
+ * @param {Function} props.onChange Callback function to be called when the texts are changed.
+ */
+export default function TextsEditor( {
+	initialTexts = [],
+	minNumberOfTexts = 0,
+	maxNumberOfTexts = 0,
+	maxCharacterCounts,
+	addButtonText,
+	placeholder,
+	onChange,
+} ) {
+	const [ texts, setTexts ] = useState( () => {
+		const shortage = minNumberOfTexts - initialTexts.length;
+		if ( shortage <= 0 ) {
+			return initialTexts;
+		}
+
+		const supplement = Array.from( { length: shortage }, () => '' );
+		return initialTexts.concat( supplement );
+	} );
+
+	const updateTexts = ( nextTexts ) => {
+		setTexts( nextTexts );
+		onChange( nextTexts );
+	};
+
+	const handleChange = ( text, { event } ) => {
+		const { index } = event.target.dataset;
+		const nextTexts = [ ...texts ];
+		nextTexts[ index ] = text.trim();
+		updateTexts( nextTexts );
+	};
+
+	const handleRemoveClick = ( event ) => {
+		const { index } = event.currentTarget.dataset;
+		const nextTexts = [ ...texts ];
+		nextTexts.splice( index, 1 );
+		updateTexts( nextTexts );
+	};
+
+	const handleAddClick = () => {
+		updateTexts( texts.concat( '' ) );
+	};
+
+	const normalizedMaxCharacterCounts = [ maxCharacterCounts ].flat();
+
+	return (
+		<div className="gla-texts-editor">
+			<div className="gla-texts-editor__text-list">
+				{ texts.map( ( text, index ) => {
+					const maxCharacterCount =
+						normalizedMaxCharacterCounts[ index ] ??
+						normalizedMaxCharacterCounts[ 0 ];
+
+					return (
+						<div
+							key={ index }
+							className="gla-texts-editor__text-item"
+						>
+							<AppInputControl
+								className="gla-texts-editor__text-input"
+								value={ text }
+								kindCharacterCount="google-ads"
```

Comment:
❓ I see is this is always going to be `google-ads`. Is this going to be extended in the future to other kinds of characters sets?

If not I would just use a boolean. With character count. Or even better just relying on `maxCharacterCount` prop for deciding it AppInputControl should care about counting characters.

Reply:
The need to specify `kindCharacterCount` was an intentional dev design. As the character count implementation of Google Ads doesn't use a common rule and the `AppInputControl` is a basic component, setting Google Ads rule as the default counter would not be ideal.

Therefore, even if no other character count rules will be added at present, it would be better to avoid misapplying Google Ads rule than to make them the default.

Reply:
I checked again but I still not get it. 

In AppInputControl we check 2 conditions to perform the new counter validation:

1- `kindCharacterCount` should be defined
2- `maxCharacterCount` > 0

So or we remove condition 1 or we could just omit `kindCharacterCount` prop and use always `google` one when condition 2 is satisfied.

Reply:
I agree that if we are going to have multiple implementations it would be necessary to add that "kind of characters" property. But since it's not even planned, I don't think it's necessary to add it.

Reply:
As mentioned earlier, this is an intentional precaution to avoid misuse. Just because the code can work well in the runtime or is not used, this does not necessarily mean that precaution is always unnecessary. It is a common practice to add precautions to error-prone places unless it's considered overengineering handling.

For example, a `getPageQuery` function could look like this:

```js
function getPageQuery( page ) {
  return {
    page,
    per_page: 10,
    order_by: 'total_sales',
    order: 'desc',
  };
}
```

It certainly works well as all previous page buttons are always disabled or hidden when viewing the first page.

However, `getPageQuery` is a common util function, and even though it's currently called without going below the minimum page number, it's still good practice to add some precautions like this:

```js
function getPageQuery( page ) {
  return {
    // Avoid `page` out of range
    page: Math.max( page, 1),
    per_page: 10,
    order_by: 'total_sales',
    order: 'desc',
  };
}
```

Another benefit is that it also shows that the minimum number of pages starts from 1.

To summarize, `'google-ads'` was intentionally not used as the default value for `kindCharacterCount` here for the following reasons.
1. It's not a common method.
2. Even among the common methods, none of them is widely considered to be the default method.
3. It's a precaution to avoid misuse.
4. Whether to set the default value for `kindCharacterCount` or not, the difference is likely to be only a few lines of code and not complicated, so it probably is not overengineering.